modules/apps/headless/headless-builder/headless-builder-impl/src/main/java/com/liferay/headless/builder/internal/resource/HeadlessBuilderResourceImpl.java

```diff
@@ -90,7 +93,16 @@ private <T> Response _executeEndpoint(
 				successUnsafeFunction)
 		throws Exception {
 
-		APIApplication.Endpoint endpoint = _endpointMatcher.getEndpoint(
+		EndpointMatcher endpointMatcher = _endpointMatcherFunction.apply(
+			_company.getCompanyId());
+
+		if (endpointMatcher == null) {
+			return Response.status(
+				Response.Status.NOT_FOUND
+			).build();
+		}
+
+		APIApplication.Endpoint endpoint = endpointMatcher.getEndpoint(
 			"/" + path);
 
 		if ((endpoint == null) || (endpoint.getScope() != scope)) {
```

Comment:
What do you think about including a new argument in the `getEndpoint` method of the EndpointMatcher for the scope?

```
APIApplication.Endpoint endpoint = endpointMatcher.getEndpoint(
	"/" + path, scope);

if (endpoint == null) {
	throw new NoSuchModelException(
		"Endpoint /%s does not exist for " + path);
}
```

I see weird to validate the scope out of the getEndpoint method :thinking: 

Furthermore, we would be avoiding the validations using the regular expression if the scope does not match:

```
_predicates.put(
	endpoint,
	(path, scope) -> {
		if (scope != endpoint.getScope()) {
			return false;
		}

		Matcher matcher = pattern.matcher(path);

		return matcher.matches();
	});
```

Reply:
Done!